linera-core/src/worker.rs

```diff
@@ -408,19 +408,19 @@ where
         certificate: Certificate,
         hashed_certificate_values: Vec<HashedCertificateValue>,
         hashed_blobs: Vec<HashedBlob>,
-        mut notifications: Option<&mut Vec<Notification>>,
+        mut notifications: Option<&mut impl Extend<Notification>>,
     ) -> Result<ChainInfoResponse, WorkerError> {
         let (response, actions) = self
             .handle_certificate(certificate, hashed_certificate_values, hashed_blobs, None)
             .await?;
-        if let Some(notifications) = notifications.as_mut() {
+        if let Some(ref mut notifications) = notifications {
```

Comment:
nit: to me `ref` is a deprecated coding style.

Reply:
I don't think that can be true — in fact the `ref mut` style is how we need to implement convenience functions like `as_mut()`!

Reply:
@Twey Are you saying implementations `as_mut()` often use `ref mut`? (Need to think about it)

FWIW they're called "legacy patterns" in the Rust book :)
https://doc.rust-lang.org/1.33.0/book/ch18-03-pattern-syntax.html#legacy-patterns-ref-and-ref-mut

> Because Rust is backwards compatible, we couldn’t remove ref and ref mut, and they’re sometimes useful in obscure situations, where you want to partially borrow part of a struct as mutable and another part as immutable. But you may see them in older Rust code, so knowing what they do is still useful.

Reply:
I think I agree with [this comment](https://users.rust-lang.org/t/is-ref-really-a-legacy-pattern/23712/6) — the ‘match ergonomics’ change results in some very spooky action at a distance, especially with large matches (which we do have in the codebase).  By contrast, the explicit `ref` syntax makes the pattern syntax less context-dependent, and decides whether the binding is by reference or by value in a very explicit way in the place where the variable is bound.  There are also still patterns you can't write without `ref`, when borrowing some fields but moving others, so calling them ‘legacy’ seems a bit disingenuous (which I assume is why the section was removed entirely from the book).  I was wrong about `as_mut()` though, you can in fact (post the ‘match ergonomics’ change) write it without `ref` — though `Option::as_mut` and friends do still use `ref`!